### DISCUSSION TITLE
[server] starting server with -e option when central server is full.

### DISCUSSION THREAD

**Discussion by jujudusud** [jamulussoftware/jamulus#1481] [Technical Support & Help]
**Describe the bug** [EDIT] -+
Starting a server with the -e option, if the central server is full: the server does not -(start)+register to this central as soon as it is full!

**To Reproduce**
- Wait until anygenre1 is full,
- start Jamulus with `-s -e anygenre1.jamulus.io:22124` options,
- if anygenre1 is selected on the list, the server -(never starts)+don't register to another central server.

**Expected behavior**
Don't know what as to be expected here, just jump automatically to another central server, maybe?

**Screenshots**
[Screenshot](https://imgur.com/ekkIrR4.png)

**Operating system**
Archlinux, but I don't know if it's OS dependant.

**Version of Jamulus**
Jamulus, Version 3.7.0
**Additional context**
Don't know if it's the same with no gui or service.

**Comment by dcorson-ticino-com** [comment]
The server will start as soon as the central server has a free slot.

**Comment by gene96817** [comment]
This intended to be a friendly question.
What is the benefit of waiting for a free slot at the central (directory) server?
Would be better for the server to start and let the registration eventually catch up? In this way, regular users of the server could use the IP address to connect. Less convenient but at least the server will be available for use.

**Comment by dcorson-ticino-com** [comment]
To be perfectly honest, I am not 100% sure that the server really doesn't start, and I didn't understand the question of the OP that way.  But it does not show up in the directory list, which for users looking for it means it isn't there, i.e. didn't start.

**Comment by jujudusud** [comment]
Oh yes ! Apologies if I don't use the right words... excuse my poor english, french guy here.
@dcorson-ticino-com you are perfectly right !
- the server starts but can't register to the central server, I will edit my first comment in order to tell the right things.
- the server will automatically register as soon as the central as a free slot. 
So, it works as expected.
yes @gene96817 it is a friendly question. :-)

Now another behavior I don't understand. When starting the server as said  (-s -e), if the central server is full and another server is selected in a prior use with the drop-down menu in jamulus server GUI, the server tries to register on the full server and after that tries the one selected (by the prior use, not intentionally) and register if it is not full. don't know if I am clear.
It seems that `-e` option filled the "custom central server address of the GUI not taking care of the drop down list "genre".

Just to discuss and understand, or tracking a bug if it is one.

**Reply by gilgongo** [reply]
>  the server tries to register on the full server and after that tries the one selected (by the prior use, not intentionally) and register if it is not full

I can't reproduce this, but I'm not sure if I'm doing it right. Do you mean:

1. Register on Directory A (with slots free)
2. Register on Directory B (with slots free)
3. Try to register on Directory C that is full
4. Server fails to register, but then automatically joins Directory B?

**Reply by jujudusud** [reply]
I start the server with the following command:
`Jamulus -s -d -e anygenre1.jamulus.io:22124 -F -o "chezju;Toulon;74" --norecord -T -u 10`

I tried again and I just saw what jamulus's behavior is.
When starting with GUI,Jamulus is not taking account of `-e` option since -e filled the "Custom Central Server Address" of the options tab and Jamulus tries to register with the value selected in the drop down menu "Genre" of the settings tabs instead of the one passed with the -e option.

Is that behaviour we wanted ?

In the [command line option](https://jamulus.io/wiki/Command-Line-Options) page of the web site, -e option is explained with the following sentence: "Makes the server public and sets its genre. See also -o".
The user starting the server with -e option certainly don't want his server registered to another one without his intervention.

I will have a look now with a non GUI server.

**Reply by gilgongo** [reply]
I'm afraid I don't understand what you are describing. 

> When starting with GUI,Jamulus is not taking account of -`e` since -e filled the "Custom Central Server Address" 

`-e` is a server option, but the  "Custom Central Server Address"  is a field in the **client** UI. Do you mean you expect the **client** to use the value you give to `-e` on the **server** as its  "Custom Central Server Address" ? In which case, no the client can't do that.

> The user starting the server with -e option certainly don't want his server registered to another one without his intervention.

Can you list the exact steps you're taking to reproduce that?

**Reply by jujudusud** [reply]
There is a `Custom central Address` in the settings tab of server window, as described in Other options of [this page of the website](https://jamulus.io/wiki/Server-Win-Mac).

You can see it in this screenshot [Image - server - Custom central Address](https://user-images.githubusercontent.com/4561747/95724775-accd3e80-0c6e-11eb-90ba-7131e9c15316.png)

Exact steps :
1. starting the server typing this command in a terminal : `Jamulus -s -d -e anygenre1.jamulus.io:22124 -F -o "chezju;Toulon;74" --norecord -T -u 10`
2. In the Jamulus server window, left clic on `Options` tab
3. Looking into `Custom Central Server Address` that the `-e` filled this.

**Comment by jujudusud** [comment]
Ok I understand what's going on.
The graphical user interface for the server is not controllable by the command line. That's it.
-e doesn't tickle the "make my server public" box either and it doesn't allow to select a custom central server from the drop-down list.
It's not a problem but I didn't read that on the web site and I think it should be explained.

**Reply by gilgongo** [reply]
So does -e only work for headless servers then?

**Reply by dcorson-ticino-com** [reply]
I just looked.  
For servers with GUI the command line options are over written by the values in the ini file, i.e. at startup they take the same state as when the server was last closed.
Headless servers do not use the ini file.

**Reply by jujudusud** [reply]
Ok nice to know.
The rule should be, "use command line to start the server with no GUI".

**Reply by dcorson-ticino-com** [reply]
>Headless servers do not use the ini file.

That is the case, a headless server i.e. no GUI, does not use the ini file.
